service/tcp_test.go

```diff
@@ -209,11 +250,220 @@ func (m *probeTestMetrics) AddUDPPacketFromTarget(clientLocation, accessKey, sta
+	}
+	return counts
+}
+
+func probe(serverAddr *net.TCPAddr, bytesToSend []byte) error {
+	conn, err := net.DialTCP("tcp", nil, serverAddr)
 	if err != nil {
-		t.Fatalf("ListenTCP failed: %v", err)
+		return fmt.Errorf("DialTCP failed: %w", err)
+	}
+
+	n, err := conn.Write(bytesToSend)
+	if err != nil || n != len(bytesToSend) {
+		return fmt.Errorf("Write failed. bytes written: %v, err: %v", n, err)
+	}
+	conn.CloseWrite()
+
+	nRead, err := conn.Read(make([]byte, 1))
+	if err != io.EOF || nRead != 0 {
+		return fmt.Errorf("Read not EOF. bytes read: %v, err: %v", nRead, err)
+	}
+	return nil
+}
+
+func TestProbeRandom(t *testing.T) {
+	listener := makeLocalhostListener(t)
+	cipherList, err := MakeTestCiphers(ss.MakeTestSecrets(1))
+	require.Nil(t, err, "MakeTestCiphers failed: %v", err)
+	testMetrics := &probeTestMetrics{}
+	s := NewTCPService(cipherList, nil, testMetrics, 200*time.Millisecond)
+	go s.Serve(listener)
+
+	buf := make([]byte, 92)
+	for numBytesToSend := 0; numBytesToSend < len(buf); numBytesToSend++ {
+		bytesToSend := buf[:numBytesToSend]
+		rand.Read(bytesToSend)
+		err := probe(listener.Addr().(*net.TCPAddr), bytesToSend)
+		require.Nil(t, err, "Failed on byte %v: %v", numBytesToSend, err)
+	}
+	s.GracefulStop()
+	require.Equal(t, len(buf), len(testMetrics.probeData))
+}
+
+func makeClientBytesBasic(t *testing.T, cipher *ss.Cipher, targetAddr string) []byte {
+	var buffer bytes.Buffer
+	socksTargetAddr := socks.ParseAddr(targetAddr)
+	require.Equal(t, 1+4+2, len(socksTargetAddr))
+	ssw := ss.NewShadowsocksWriter(&buffer, cipher)
+	n, err := ssw.Write(socksTargetAddr)
+	require.Nil(t, err, "Write failed: %v", err)
+	require.Equal(t, len(socksTargetAddr), n, "Write failed: %v", err)
+	require.Equal(t, 73, buffer.Len())
+
+	n, err = ssw.Write([]byte("initial data"))
+	require.Nil(t, err, "Write failed: %v", err)
+	require.Equal(t, 73+2+16+12+16, buffer.Len()) // 119
+
+	n, err = ssw.Write([]byte("more data"))
+	require.Nil(t, err, "Write failed: %v", err)
+	require.Equal(t, 119+2+16+9+16, buffer.Len()) // 162
+
+	return buffer.Bytes()
+}
+
+func makeClientBytesCoalesced(t *testing.T, cipher *ss.Cipher, targetAddr string) []byte {
+	var buffer bytes.Buffer
+	socksTargetAddr := socks.ParseAddr(targetAddr)
+	ssw := ss.NewShadowsocksWriter(&buffer, cipher)
+	n, err := ssw.LazyWrite(socksTargetAddr)
+	require.Nil(t, err, "LazyWrite failed: %v", err)
+	require.Equal(t, len(socksTargetAddr), n, "LazyWrite failed: %v", err)
+	n, err = ssw.Write([]byte("initial data"))
+	require.Nil(t, err, "Write failed: %v", err)
+	require.Equal(t, 32+2+16+7+12+16, buffer.Len()) // 85
+
+	n, err = ssw.Write([]byte("more data"))
+	require.Nil(t, err, "Write failed: %v", err)
+	return buffer.Bytes()
+}
+
+func firstCipher(cipherList CipherList) *ss.Cipher {
+	snapshot := cipherList.SnapshotForClientIP(nil)
+	cipherEntry := snapshot[0].Value.(*CipherEntry)
+	return cipherEntry.Cipher
+}
+
+func TestProbeClientBytesBasicTruncated(t *testing.T) {
```

Comment:
It seems like there's a lot of duplicate code for these tests.  I think it would be worth factoring out the test execution logic, so each test just has to choose the input data, call the shared execution function, and check the status counts.

Reply:
There isn't actually a lot of duplicated code. A lot of it is the probe payload generation and the assertions, which differ. And the set up code needs to be split so I have access to the cipher, the discard server and the metrics, which is annoying. Here is an example of trying to refactor the server creation: 
![image](https://user-images.githubusercontent.com/113565/96957676-4693b780-14c9-11eb-9b45-faddbc4bf6e5.png)

Reply:
OK